src/garage/tf/algos/npo2.py

```diff
@@ -0,0 +1,575 @@
+
+    @overrides
+    def init_opt(self):
+        """Initialize optimizater."""
+        pol_loss_inputs, pol_opt_inputs = self._build_inputs()
+        self._policy_opt_inputs = pol_opt_inputs
+
+        pol_loss, pol_kl = self._build_policy_loss(pol_loss_inputs)
+        self.optimizer.update_opt(loss=pol_loss,
+                                  target=self.policy,
+                                  leq_constraint=(pol_kl, self.max_kl_step),
+                                  inputs=flatten_inputs(
+                                      self._policy_opt_inputs),
+                                  constraint_name='mean_kl')
+
+        return dict()
+
+    @overrides
+    def optimize_policy(self, itr, samples_data):
+        """Optimize policy."""
+        policy_opt_input_values = self._policy_opt_input_values(samples_data)
+        # Train policy network
+        logger.log('Computing loss before')
+        loss_before = self.optimizer.loss(policy_opt_input_values)
+        logger.log('Computing KL before')
+        policy_kl_before = self.f_policy_kl(*policy_opt_input_values)
+        logger.log('Optimizing')
+        self.optimizer.optimize(policy_opt_input_values)
+        logger.log('Computing KL after')
+        policy_kl = self.f_policy_kl(*policy_opt_input_values)
+        logger.log('Computing loss after')
+        loss_after = self.optimizer.loss(policy_opt_input_values)
+        tabular.record('{}/LossBefore'.format(self.policy.name), loss_before)
+        tabular.record('{}/LossAfter'.format(self.policy.name), loss_after)
+        tabular.record('{}/dLoss'.format(self.policy.name),
+                       loss_before - loss_after)
+        tabular.record('{}/KLBefore'.format(self.policy.name),
+                       policy_kl_before)
+        tabular.record('{}/KL'.format(self.policy.name), policy_kl)
+        pol_ent = self.f_policy_entropy(*policy_opt_input_values)
+        tabular.record('{}/Entropy'.format(self.policy.name), np.mean(pol_ent))
+
+        self._fit_baseline(samples_data)
+        self._policy_update_ops()
+
+    @overrides
+    def get_itr_snapshot(self, itr):
+        """Get iteration snapshot."""
+        return dict(
+            itr=itr,
+            policy=self.policy,
+            baseline=self.baseline,
+        )
+
+    def _build_inputs(self):
+        action_space = self.policy.action_space
+
+        with tf.name_scope('inputs'):
+            obs_var = self.policy.model.networks['default'].input
+            old_obs_var = self.old_policy.model.networks['default'].input
+            # Ops to assign new policy weights to old policy
+            target_update_op = tensor_utils.get_target_ops(
+                self.policy.get_global_vars(),
+                self.old_policy.get_global_vars())
+            self._policy_update_ops = tf.get_default_session().make_callable(
+                target_update_op)
+
+            action_var = action_space.to_tf_placeholder(name='action',
```

Comment:
there are 4 different ways of creating a tensor here. can we deleted 2 or 3 of them?